### PR TITLE
Fix softmax bug for ragged tensors

### PR DESCRIPTION
**Pull request by maresb**
The current implementation of softmax for ragged tensors is numerically unstable. In particular, the `exp()` function is computed on the raw logits. This causes an overflow when the logits are large positive. When the logits are large negative, this leads to an underflow and division by zero. The problem manifests through `nan` entries in the ragged tensor.

In this PR I (hopefully) fix the ragged tensor implementation and implement a corresponding test.

For reference of a numerically stable implementation, consider the [numpy implementation](https://github.com/maresb/tensorflow/blob/8e75354bae33d10bd43091534ed5e1c1062d85db/tensorflow/python/ops/nn_test.py#L97-L104) in the test suite.

(Recall that the softmax function is invariant under adding an overall constant to the input logits. Before applying `exp()` it is advantageous to subtract the max so that one logit component is zero and the rest are negative. This ensures that the denominator is numerically finite.)

### CODE REVIEW

**Review by bishalbaaniya123 — APPROVED** [review]

**Comment by maresb** [comment]
Oops, I seem to have messed up something. :frowning_face: 

```
Traceback (most recent call last):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/nn_test_cpu.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 1493, in decorated
    f(self, *args, **kwargs)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/nn_test_cpu.runfiles/org_tensorflow/tensorflow/python/ops/nn_test.py", line 164, in testRagged
    self.assertAllClose(y_tf, expected, eps)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/nn_test_cpu.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 1533, in decorated
    return f(*args, **kwds)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/nn_test_cpu.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 3108, in assertAllClose
    self._assertAllCloseRecursive(a, b, rtol=rtol, atol=atol, msg=msg)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/nn_test_cpu.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 3012, in _assertAllCloseRecursive
    return self._assertRaggedClose(a, b, rtol, atol, msg)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/nn_test_cpu.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 3596, in _assertRaggedClose
    self._assertListCloseRecursive(a_list, b_list, rtol, atol, msg)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/nn_test_cpu.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 3608, in _assertListCloseRecursive
    self._assertListCloseRecursive(a[i], b[i], rtol, atol, msg,
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/nn_test_cpu.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 3606, in _assertListCloseRecursive
    self.assertLen(a, len(b), "Length differs for %s" % path)
AssertionError: [0.6652409434318542, 0.2447284758090973, 0.09003057330846786] has length of 3, expected 1. : Length differs for value[0]
```

**Comment by maresb** [comment]
Ok, it seems to be succeeding now. 😅

**Comment by gbaned** [comment]
@rohan100jain  Can you please review this PR ? Thank you!

File: tensorflow/python/ops/nn_test.py [hunk 1]
```diff
@@ -151,6 +153,18 @@ def testGradient(self, x_shape):
     eps = 2e-8
     self.assertLess(err, eps)
 
+  @test_util.run_in_graph_and_eager_modes
+  def testRagged(self):
+    x_list = [[-500, -501, -502], [], [1729, 1729]]
```
Comment (cantonios) [inline tensorflow/python/ops/nn_test.py:158]: Add a few more relevant test cases here (including some "regular" ones).

**Review by cantonios — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/nn_test.py [hunk 2]
```diff
@@ -151,6 +153,18 @@ def testGradient(self, x_shape):
     eps = 2e-8
     self.assertLess(err, eps)
 
+  @test_util.run_in_graph_and_eager_modes
+  def testRagged(self):
```
[the remark this answers is not in the document]
Comment (maresb) [reply tensorflow/python/ops/nn_test.py]: Done (under the assumption that a migration of all the other ragged ops tests into this new file is out-of-scope for this PR)

File: tensorflow/python/ops/nn_test.py (hunk 1 above)
Comment (maresb) [reply tensorflow/python/ops/nn_test.py]: Ok, I added a test case with "ordinary values", and split this into cases with "large values" and "short tensors" (a ragged tensor where lengths 0 and 1 occur).

**Comment by maresb** [comment]
@cantonios, thanks so much for the review! Have I satisfactorily fixed everything?

**Review by cantonios — APPROVED** [review]

**Comment by cantonios** [comment]
I think so - we'll let it run through the whole test suite.  If everything looks good, it will be merged.

**Comment by maresb** [comment]
@cantonios the test Ubuntu GPU, Python 3.9 is failing. I found a few other recent PRs where it seems to be succeeding, so I'm wondering if my PR is to blame?

I was looking through what I could access from the logs, and I didn't manage to find any sort of explanation for the failure such as a stack trace. Am I missing something, or is that info only Google-accessible?

I found `kokoro-sisyphus_resultstore_prod_tensorflow_rel_docker_github_presubmit_pycpp_gpu_py39_dfdf902b-73bc-4129-a7bd-541ec0226cb2_build.log`, which contains:

```
All these execution steps failed:
bazel_test
You can find truncated logs under the test target. All the logs are
also combined in the Build Log (above) for this invocation.

If there were test failures, you can find a deduplicated list of all of them
under Targets > pkg/pip_and_nonpip_tests.
This test uses Docker. See go/tf-devinfra/docker for help.


[ID: 9864512] Build finished after 1798 secs, exit value: 1
```

**Comment by cantonios** [comment]
I think a test or two timed out.  Doesn't look related.

**Comment by maresb** [comment]
Is there anything blocking a merge here? (It'd be nice to merge while this is fresh and before any conflicts arise, but I understand if there's a release cycle or if the team has other priorities.) Let me know if I can help with anything. Thanks so much!!!

**Pull request merged**